core/arch/arm/mm/tee_mmu.c

```diff
@@ -74,84 +66,86 @@
 /* Support for 31 concurrent sessions */
 static uint32_t g_asid = 0xffffffff;
 
-static void tee_mmu_umap_set_pa(struct tee_mmap_region *tbl,
-			size_t granule, paddr_t pa, size_t size, uint32_t attr)
+static TEE_Result tee_mmu_umap_add_param(struct tee_mmu_info *mmu,
+					 struct param_mem *mem)
 {
-	paddr_t upa = ROUNDDOWN(pa, granule);
-	size_t usz = ROUNDUP(pa - upa + size, granule);
-
-	tbl->pa = upa;
-	tbl->size = usz;
-	tbl->attr = attr;
-}
-
-static TEE_Result tee_mmu_umap_add_param(struct tee_mmu_info *mmu, paddr_t pa,
-			size_t size, uint32_t attr)
-{
-	struct tee_mmap_region *last_entry = NULL;
+	TEE_Result res;
+	struct tee_ta_region *last_entry = NULL;
 	size_t n;
-	paddr_t npa;
+	uint32_t cattr;
+	uint32_t attr;
 	size_t nsz;
+	size_t noffs;
+
+	res = mobj_get_cattr(mem->mobj, &cattr);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	attr = cattr << TEE_MATTR_CACHE_SHIFT;
+	attr |= TEE_MMU_UDATA_ATTR;
+	if (!mobj_is_secure(mem->mobj))
+		attr &= ~TEE_MATTR_SECURE;
 
 	/* Check that we can map memory using this attribute */
 	if (!core_mmu_mattr_is_ok(attr))
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	/* Find empty entry */
 	for (n = TEE_MMU_UMAP_PARAM_IDX; n < TEE_MMU_UMAP_MAX_ENTRIES; n++)
-		if (!mmu->table[n].size)
+		if (!mmu->regions[n].size)
 			break;
 
 	if (n == TEE_MMU_UMAP_MAX_ENTRIES) {
 		/* No entries left "can't happen" */
 		return TEE_ERROR_EXCESS_DATA;
 	}
 
-	tee_mmu_umap_set_pa(mmu->table + n, CORE_MMU_USER_PARAM_SIZE,
-			    pa, size, attr);
+	mmu->regions[n].mobj = mem->mobj;
```

Comment:
this assumes `mem->mobj` is aligned on block/page boundaries as later `mobj_get_pa()` will be used to setup the mmu entry (in `core_mmu_populate_user_map()`)